MiniShopping/MiniShopping/Models/FetchProducts.m

```diff
@@ -0,0 +1,43 @@
+#import "FetchProducts.h"
+#import "ImageDownloader.h"
+
+@implementation FetchProducts
+
+- (void)fetchData {
+    NSString *urlString = @"https://dummyjson.com/products";
+    NSURL *url = [NSURL URLWithString:urlString];
+    
+    NSURLSessionTask *task = [[NSURLSession sharedSession] dataTaskWithURL:url completionHandler:^(NSData * _Nullable data, NSURLResponse * _Nullable response, NSError * _Nullable error) {
+        if (error) {
+            NSLog(@"URL Session Error :  %@", error);
+            return;
+        }
+        if (!data) {
+            NSLog(@"Data Error :  %@", error);
+            return;
+        }
+        NSError *jsonError;
+        NSDictionary *jsonData = [NSJSONSerialization JSONObjectWithData:data options:kNilOptions error:&jsonError];
+        if (jsonError) {
+            NSLog(@"JSON Error : %@", jsonError);
+            return;
+        }
+        NSArray *productsArray = jsonData[@"products"];
+        for (NSDictionary *productDict in productsArray) {
+            Product *product = [[Product alloc] initWithDictionary:productDict];
+            NSString *thumbnailURLString = productDict[@"thumbnail"];
+            ImageDownloader *imageDownloader = [[ImageDownloader alloc] init];
+            [imageDownloader downloadImageWithURL:thumbnailURLString completion:^(UIImage *image, NSError *error) {
+                if (error) {
+                    NSLog(@"Image Download Error %@", error);
+                    return;
+                }
+                [product setThumbnailImage:image];
+            }];
+            [self.delegate addToProductsArray:product];
```

Comment:
Product images should only be fetched for visible cells on the screens. With this current code, it will try to fetch all products image which will unnecessarily consumes Network data/battery, resources etc. 

We should fetch resources on-demand to improve the performance.